Add ERC: Structured Data Clear Signing Format (#509)
ethereum/ERCs: ERCS/erc-7730.md

---
eip: 7730
title: Structured Data Clear Signing Format
description: Draft specification of a json format providing additional description of how to clear-sign smartcontract calls and typed messages
author: Laurent Castillo (@lcastillo-ledger)
discussions-to: https://ethereum-magicians.org/t/eip-7730-proposal-for-a-clear-signing-standard-format-for-wallets/20403
status: Draft
type: Standards Track
category: ERC
created: 2024-02-07

> requires: 712

Reviewer: ```suggestion
requires: 155, 712
```

Providing this additional formatting information requires deep knowledge of the way a smart contract or message is going to be used. It is expected that DApp developers will be the best placed to write such a file. The intent of an open standard is to only write this file once and have it work with most wallets supporting this standard.

> *Deployment model*
> 
> Making the ERC-7730 available for wallets is a key factor of adoption. We have a few options:
> - In each dApps github repository or web site: good for autonomy but can be problematic for discoverability by wallets of new ERC-7730 files
> - Foundation operated repository, like ethereum chainID list: good alternative between decentralization and discoverability.
> - Ledger repository: as a short term solution, Ledger is providing a central repository (See Ledger github repository)

## Specification

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.

> ### Simple example

Reviewer: ```suggestion
### Example
```

I would just go with "Example" because simple is subjective :rofl:

The following is a simple example of how to clear sign a `transfer` function call on an [ERC-20](./eip-20.md) contract.

```json
{

>     "$schema": "https://github.com/LedgerHQ/clear-signing-erc7730-registry/blob/master/specs/erc7730-v1.schema.json",

Reviewer: This should be a file in your proposal's `assets/` directory.

                {
                    "chainId": 137,
                    "address": "0xc2132D05D31c914a87C6611C10748AEb04B58e8F",
                },
                {
                    "chainId": 42161,
                    "address": "0xFd086bC7CD5C481DCC9C85ebE478A1C0b69FCbb9",
                }
            ]
        }
    }, 

    "metadata": {
        "owner": "Tether",
        "info": {
            "legalName": "Tether Limited",
            "url": "https://tether.to/",
            "lastUpdate": "2017-11-28T12:41:21Z"  
        }
    },

    "display": {
        "formats": {
            "transfer(address,uint256)": {
                "intent": "Send",
                "fields": {
                    "_to" : {
                        "label": "To",
                        "format": "addressOrName"
                    },
                    "_value" : {
                        "label": "Amount",
                        "format": "tokenAmount",
                        "params": {
                            "tokenPath": "@.to"
                        }
                    }
                },
                "required": ["_to", "_value"]
            }
        }
    }
}
```

> The `$schema` key refers to the latest version of this specification json schema (currently version 1).

Reviewer: > (currently version 1)

Once this proposal goes final it can no longer be updated, so when you publish version 2 this text would become misleading. One option is:

```suggestion
The `$schema` key refers to the latest version of this specification json schema (version 1 at the time of writing).
```

* A description of the type system, the *schema*, that should allow splitting the data into *fields*, each field clearly identified with a *path* that can be descrived as a string.

Displaying structured data is often done by wallets to review its content before authorizing an action in the form of a *signature* over some serialization of the structured data. As such, the structured data is contained in a *container structure*:
* Container structure has a well defined *signature* scheme (a serialization scheme, a hashing scheme, and signature algorithm).
* The container structure does not necessarily follow the same type system as the structured data.
* Wallets receive the full container structure and uses the signature scheme to generate a signature on the overall structure.

```mermaid
block-beta
  columns 1
  block:container
    columns 3
    title["Container Structure"] space:2
    a["Data"]
    struct["Structured Data"]
    c["Data"]
  end
  blockArrowId4<["Wallet"]>(down)
  block:signature
    columns 3
    space sig["Signature"] space
  end
  style title fill:transparent,stroke:transparent
  style struct fill:#555

> ```

Reviewer: Mermaid is not rendered by our tooling. Please replace this with an image in your `assets/` directory. We prefer SVG, but PNG is also acceptable.

Current specification covers EVM smart contract calldata:
* Defined in Solidity
* The schema is the ABI (expected in json format when linked to))
* The container structure is an EVM Transaction serialized in RLP encoding

It also supports EIP-712 messages
* Defined in EIP-712
* The schema is extracted from the message itself, from the `types` definitions and the `primaryType` key.
* An EIP-712 message is self contained, the signature is applied to the hashed message itself.
  

> In the future, it could be extended to structured data like Meta Transaction in [EIP-2771](./eip-2771.md) or User Operations in [EIP-4337](./eip-4337.md).

Reviewer: ```suggestion
```

If you'd like to talk about design choices made for future extensibility, do so in the Rationale section.

* `#.params.path[0..19]` refers to the first 20 bytes of the `path` byte array
* `#.params.path[-20:-1]` refers to the last 20 bytes of the `path` byte array

References to values in the format specification file
* `$.metadata.enums.interestRateMode` refers to the values of an enum defined in the specification file (typically to convert an integer into a displayed string)
* `$.display.definitions.minReceiveAmount` refers to a common definition reused accross fields formatting definition

References to values in the container (here a EVM Tx container)
* `@.amount` refers to the enclosing transaction native currency amount
* `@.to` refers to the enclosing transaction destination address (usually, a smartcontract bound to this ERC-7730 through the `context` section)

#### Organizing files

Smart contracts and EIP-712 messages are often re-using common interfaces or types that share similar display formatting. This specification supports a basic inclusion mechanism that enables sharing files describing specific interfaces or types.

> The `includes` top-level key is an URL pointing to an ERC-7730 file that MUST follow this specification.

Reviewer: ```suggestion
The `includes` top-level key is a URL pointing to an ERC-7730 file that MUST follow this specification.
```

  * `address`: the address of the deployed contract on specified `chainId` chain.

---
*The following constraints for contracts are considered draft*

**`contract.addressMatcher`**

An URL that can be used to check whether the current transaction `(chainId, contract address)` can be formatted using the ERC-7730 file.

A wallet MAY use this URL to check unknown addresses, default is to fail constraint on unknown addresses.

**`contract.factory`**

An object describing the factory used to deploy smartcontracts that can be clear signed using the ERC-7730 file.

A factory is a json object with:
* `deployEvent` key, specifying the solidity signature of the events emitted when deploying a clear-signable contract.
* `deployments` key: an array of deployment options as in `contract.deployments`. These deployments represent the addresses at which the *factory* is deployed.  

To verify a factory constraints a wallet MUST check that:
* The current transaction destination address is included in an event of signature `factory.deployEvent`
* The emitter of the event is a factory contract deployed at an address matching one of the deployment option in `factory.deployments`
---

*Examples*

> [TBD]

Reviewer: If you use HTML-style comments, the linter will make sure you replace them before moving to Review:

```suggestion
<!-- TBD -->
```

#### Address trusted sources

Address names trusted sources specify which type and source of trusted names SHOULD be used to replace an address with a human readable names. 

When specified a wallet MUST only use specified sources to resolve address names. Wallet MUST verify the type of the address if able to. 

When omitted, a wallet MAY use any source to resolve an address.

| Address type  | Description |
| ---           | --- |
| wallet        | Address is an account controlled by the wallet |
| eoa           | Address is an Externally Owned Account         |
| contract      | Address is a well known smartcontract  |
| token         | Address is a well known ERC-20 token  |
| collection    | Address is a well known NFT collection  |

A wallet MAY verify that a `wallet` address is in fact controlled by the wallet, and reject signing if not the case.

| Source type   | Description |
| ---           | --- |
| local         | Address MAY be replaced with a local name trusted by user. Wallets MAY consider that `local` setting for `sources` is always valid |
| ens           | Address MAY be replaced with an associated ENS domain |

### Wallets

#### Ledger

> [TBD] Link to ledger specification

Reviewer: ```suggestion
<!-- TBD --> Link to ledger specification
```

Just a note, most external links aren't allowed. If you believe a particular origin should be allowed, please see [EIP-5757](https://eips.ethereum.org/EIPS/eip-5757).

* In addition, a set of edition tools will be created and open sourced to ease visualization of the results for end users

### Wallet limitations

Wide support by wallets is key for adoption of this specification. 

Hardware wallets tend to have more limited capabilities that will impact what they can display securely, especially since the intention of the specification is to create a strong security binding between the spec and the data being reviewed.

This consideration is driving a few choices done for ERC-7730:
* Complex UI constructs like layouts, grouping and re-ordering of fields have been left to a wallet specific section, yet unspecified. After a time, we may see patterns emerge between wallets in term of minimal features.
* Representation of fields has been created to allow "flattening" the list of fields, when handling complex message structures. This flattened representation is expected to work better with Hardware wallets in particular, and is recommended at first.
* Some formatters that require recursive constructs, like `calldata` are expected to work with restrictions at first, especially on Hardware wallets.  

## Test Cases

> More examples can be found on the Ledger ERC-7730 public repository

Reviewer: This is an external link. Please remove it, or move the content into your `assets/` directory.

### Why is the ABI still necessary? 

The full signature of a function call does not contain information about the types of complex parameters that might be used in the function call, only their names, while the ABI does carry this information. This is why the ABI is the reference and the function signature MUST match it. The full form of the function call is allowed in order to simplify writing and reading the ERC-7730 file. 

### Deployment model
 
Making the ERC-7730 available for wallets is a key factor of adoption. We have a few options:
- In each dApps GitHub repository or website: good for autonomy but can be problematic for discoverability by wallets of new ERC-7730 files
- Foundation operated repository, like ethereum chainID list: good alternative between decentralization and discoverability.
- Ledger repository: as a short term solution, Ledger is providing a central repository (See Ledger GitHub repository)

### Extensibility to other Structured Data formats

In the future, this specification could be extended to structured data like Meta Transaction in [EIP-2771](./eip-2771.md) or User Operations in [EIP-4337](./eip-4337.md).

## Test Cases

> More examples can be found in the asset folder.

Reviewer: You should link to the asset.

## Security Considerations

The main security concern introduced by ERC-7730 is to avoid attacks that would use the ERC-7730 formatting mechanism to trick users into signing something wrong.

Two main attack vectors that needs to be mitigated:
* Avoid using a well-formed, trusted ERC-7730 file on data that should not be formatted using it
* Avoid badly formed, incorrect ERC-7730 files that would hide critical parameters or create misinterpretations through bed formatting

### Binding context

The binding `context` is the way we mitigate the first attack, by specifying in the ERC-7730 file itself, **what kind** of structured data should be formatted by said file.

> dApps developers MUST make sure that constraints in the binding context are restrictive enough to apply only to relevant data.

Reviewer: You should keep requirements (the all caps keywords) to the specification section.

---
eip: 7730
title: Structured Data Clear Signing Format
description: Draft specification of a json format providing additional description of how to clear-sign smartcontract calls and EIP712 messages on any wallets
author: Laurent Castillo (@lcastillo-ledger)

> discussions-to: https://github.com/LedgerHQ/clear-signing-format-spec/issues

Reviewer: This must be a thread on https://ethereum-magicians.org/.

status: Draft
type: Standards Track
category: Interface
created: 2024-02-07
requires: 712
---

## Abstract

> This specification defines a JSON format carrying additional information required to correctly display structured data to a human for review on a wallet screen, before signature by the wallet. In the current scope, structured data supported are smart contract calls in EVM transactions and EIP 712 structured messages. 

Reviewer: It would be nice to include a summary of how the proposal works. For example, what additional information is carried?

After reading your Abstract, your audience should have a high-level (but still technical) idea of what your proposal does.

        "$id": "Tether USD",
        "contract" : {
            "chainId": 1,
            "address": "0xdAC17F958D2ee523a2206206994597C13D831ec7",
            "abi": "https://api.etherscan.io/api?module=contract&action=getabi&address=0xdac17f958d2ee523a2206206994597c13d831ec7"
        }
    }, 

    "metadata": {
        "owner": "Tether",
        "info": {
            "legalName": "Tether Limited",
            "url": "https://tether.to/",
            "lastUpdate": "2017-11-28T12:41:21Z"  
        }
    },

    "display": {
        "formats": {
            "0xa9059cbb": {
                "$id" : "transfer",
                "intent": "Send",
                "fields": {
                    "_to" : {
                        "label": "To",
                        "format": "addressOrName"
                    },
                    "_value" : {
                        "label": "Amount",
                        "format": "tokenAmount",
                        "params": {
                            "tokenPath": "$.context.contract.address"
                        }
                    }
                },
                "required": ["_to", "_value"]
            }
        }
    }
}
```

> The `$schema` key refers to the latest version of this specification json schema (currently version XXX).

Reviewer: ```suggestion
The `$schema` key refers to the latest version of this specification json schema (currently version XXX <!-- TODO -->).
```

If you use HTML-style comments, the linter will ensure you replace them before going into Review.

### Common concepts

#### Structured Data

This specification intends to be extensible to describe the display formatting of any kind of structured data. 

By "Structured data", we target any data format that has:
* A well defined *type* system, the data being described itself being of a specific type
* A description of the type system, the *schema*, that should allow splitting the data into *fields*, each field clearly identified with a *path* that can be descrived as a string.
* Structured data can be part of of *container structure*, or be self-contained.
  * Container structure has a well defined *signature* scheme (either a serialization scheme or a hashing scheme, and signature algorithm).
  * The container structure does not necessarily follow the same type system as the structured data.
  * Wallets receive the container structure and uses the signature scheme to generate a signature on the overall structure.

[Show a diagram of internals of a Tx]

Current specification covers EVM smart contract calldata:

> * Defined in [Solidity](https://docs.soliditylang.org/en/latest/abi-spec.html#)

Reviewer: You'll have to remove this external link.

---
eip: 7730
title: Structured Data Clear Signing Format
description: Draft specification of a json format providing additional
  description of how to clear-sign smartcontract calls and EIP712 messages on
  any wallets

> author: Laurent Castillo(@lcastillo-ledger)

Reviewer 2: ```suggestion
author: Laurent Castillo (@lcastillo-ledger)
```

discussions-to: https://github.com/LedgerHQ/clear-signing-format-spec/issues
status: Draft
type: Standards Track
category: Interface
created: 2024-02-07
requires: 712
---

## Abstract

This specification defines a JSON format carrying additional information required to correctly display structured data to a human for review on a wallet screen, before signature by the wallet. In the current scope, structured data supported are smart contract calls in EVM transactions and EIP 712 structured messages. 

## Motivation

> Properly validating a transaction on a Hardware Wallets screen (also known as Clear Signing) is a key element of good security practices for end users when interacting with any Blockchain. Unfortunately, most data to sign, even enriched with the data structure description (ABI or EIP712 types) are not self-sufficient in term of correctly displaying them to the user for review. Among other things:

Reviewer 2: ```suggestion
Properly validating a transaction on a Hardware Wallets screen (also known as Clear Signing) is a key element of good security practices for end users when interacting with any Blockchain. Unfortunately, most data to sign, even enriched with the data structure description (ABI or EIP-712 types) are not self-sufficient in term of correctly displaying them to the user for review. Among other things:
```

> - In each dapps repository (discoverability?)
> - Foundation operated repository
> - Ledger repository

## Specification

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.

### Simple example

The following is a simple example of how to clear sign a `transfer` function call on an ERC20 contract.

```json
{
    "$schema": "https://github.com/LedgerHQ/clear-signing-format-spec/blob/master/schemas/clear-signing-schema.json",

    "context": {
        "$id": "Tether USD",
        "contract" : {
            "chainId": 1,
            "address": "0xdAC17F958D2ee523a2206206994597C13D831ec7",
            "abi": "https://api.etherscan.io/api?module=contract&action=getabi&address=0xdac17f958d2ee523a2206206994597c13d831ec7"
        }
    }, 

    "metadata": {
        "owner": "Tether",
        "info": {
            "legalName": "Tether Limited",
            "url": "https://tether.to/",
            "lastUpdate": "2017-11-28T12:41:21Z"  
        }
    },

    "display": {
        "formats": {

>             "0xa9059cbb": {

Reviewer 3: Is there a benefit in using the 4-byte signature as opposed to using the function name i.e. `transfer` or `transfer(address,uint256)`?

Author: You are right that using the function signature would be more readable! We can't use just the function name since we need unicity of those keys, but the full signature used for selector computation would work. Will do an update to allow for both selectors and function signatures

                "$id" : "transfer",
                "intent": "Send",
                "fields": {
                    "_to" : {
                        "label": "To",
                        "format": "addressOrName"
                    },

>                     "_value" : {

Reviewer 3: Some contracts that implement the same standard like ERC-20 (and should ideally be using the same specification) are using different field names e.g. in USDT they are called `_to` and `_value` while in the DAI contract they are called `dst` and `wad`. Could this cause any issues when trying to parse the ABI? Also, is it expected that the there will be multiple specifications for the same function signature?

Author: That's why the 7730 file contains a reference of the ABI used. Field name should be the same as the ones in the ABI provided as reference. Since parameter name are not part of function signatures, that shouldn't prevent reusing the erc7730 file for other contracts that use different names for parameters

### Common concepts

#### Structured Data

This specification intends to be extensible to describe the display formatting of any kind of structured data. 

By "Structured data", we target any data format that has:
* A well defined *type* system, the data being described itself being of a specific type
* A description of the type system, the *schema*, that should allow splitting the data into *fields*, each field clearly identified with a *path* that can be descrived as a string.
* Structured data can be part of of *container structure*, or be self-contained.
  * Container structure has a well defined *signature* scheme (either a serialization scheme or a hashing scheme, and signature algorithm).
  * The container structure does not necessarily follow the same type system as the structured data.
  * Wallets receive the container structure and uses the signature scheme to generate a signature on the overall structure.

[Show a diagram of internals of a Tx]

Current specification covers EVM smart contract calldata:
* Defined in [Solidity](https://docs.soliditylang.org/en/latest/abi-spec.html#)
* The ABI is the schema
* The container structure is an EVM Transaction

> It also supports EIP 712 messages

Reviewer 2: ```suggestion
It also supports EIP-712 messages
```

            "0x095ea7b3": {
                "$id" : "approve",
                "intent": "Approve",
                "fields": {
                    "_spender" : {
                        "label": "Spender",
                        "format": "addressOrName"
                    },
                    "_value" : {
                        "label": "Amount",
                        "format": "allowanceAmount",
                        "params": {
                            "tokenPath": "$.context.contract.address",
                            "threshold": "0x8000000000000000000000000000000000000000000000000000000000000000"
                        }
                    }
                },
                "required": ["_spender", "_value"]
            }

        }
    }
}
```
Note that there are no keys for binding the contract to a specific address or owner, nor any contract specific metadata.

The following file would include this generic interface and bind it to the specific USDT contract:
```json
{
    "context": {
        "$id": "Tether USD",
        "contract" : {
            "chainId": 1,
            "address": "0xdAC17F958D2ee523a2206206994597C13D831ec7"
        }
    }, 

>     "includes": ["../examples/clear-signing-example-erc20.json"],

Reviewer 3: Allowing to include more than one file per context adds complexity when merging. Is this necessary or could it be limited to a single file include per context? What would be a use-case of including more than one file?

Author: Didn't find a good example that would require more than one include for now, so definitely could be limited until we have concrete examples when it would be useful

    "metadata": {
        "owner": "Tether",
        "info": {
            "legalName": "Tether Limited",
            "url": "https://tether.to/",
            "lastUpdate": "2017-11-28T12:41:21Z"  
        },
        "token": {
            "ticker": "USDT",
            "name": "Tether USD",
            "magnitude": 6
        }
    }
}
```
Note that the keys under `context.contract` would be merged together to construct a full contract binding object.

### Context section

The `context` section describes the context to which this display format description file is bound. A wallet MUST verify that the structured data and container it is trying to sign corresponds to the context in this section.

> The current version of this specification only supports two types of binding context, EVM smartcontracts and EIP712 domains. 

Reviewer 3: ```suggestion
The current version of this specification only supports two types of binding context, EVM smart contracts and EIP712 domains. 
```

* The `$id` sub-key is an internal identifier, not used for formatting the display, only relevant to provide a human readable name to this specification file
* The `contract` sub-key is used to describe an EVM smartcontract binding context, with:
  

>   * `chainId`: an [EIP 155 identifier](https://chainid.network/) of the chain the contract described is deployed on

Reviewer 3: ```suggestion
  * `chainId`: an [EIP-155 identifier](https://chainid.network/) of the chain the descibed contract is deployed at
```

* Each field format might have parameters in a `params` sub-key. Available parameters are descried in the [Reference](#field-formats) section
* Each field definition can have an optional internal `$id` used to identify easily which field is described 

*Examples*

## Reference

### Container structure values

This section describes all container structure supported by this specification and possible references path to relevant values.

#### EVM Transaction container

| Value reference | Value Description | Examples |
| --- | --- | --- |
| @.value        | The native currency value of the transaction containing the structured data | |

### Field formats

#### Integer formats

Formats useable for uint/int solidity types 

| Format name | Parameters | Format Description | Examples |
| ----------- | ---------- | ------------------ | -------- |
| raw         | n/a        | Raw int value      | Value 1000 displayed as `1000` |
| amount      | n/a        | Display as an amount in native currency, using best ticker / magnitude match.    | Value 0x2c1c986f1c48000is displayed as `0.19866144 Eth` |

> | tokenAmount | tokenPath  | Convert value using magnitude of token, and append ticker name.    | Value 1000000, for ticker DAI, magnitude 6 displayed as `1 DAI` |

Reviewer 3: Is magnitude the same as decimals? I think the latter is used a bit more commonly.

Author: Indeed, same, will change the name
